Project 03 Peer Review PR

mcspleen opened the pull request:
Group member for this project is @Tiange

mcspleen committed:
Project 03 Peer Review PR

Group member for this project is @Tiange

mcspleen committed:
Add files via upload

mcspleen committed:
Add files via upload

mcspleen committed:
Delete project03/2025-09-28_TF(1).Rmd

Tiange committed:
Update README.md

Tiange committed:
Update project03.Rmd

Cleaned and updated some contents

mcspleen committed:
Add files via upload

mcspleen committed:
Update project03.Rmd

mcspleen committed:
Update README.md

mcspleen committed:
Merge branch 'project03_PR' into project03_PR

mcspleen committed:
Merge pull request #2 from Tiange/project03_PR

Project03 update: readme file and code

mcspleen committed:
Update README.md

mcspleen committed:
Update README.md

mcspleen committed:
Update README.md

e12chaos commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,189 @@
+# Here we test your Gibbs sampler.
+# You do not need to edit this or the section below. This is the Driver program
+
+#read promoters, store in a list of strings
+dir_wk = "/courses/BINF6250.202610/students/feng.tiang/project_03"
+folder_seq = "data/GCF_000009045.1_ASM904v1_genomic.fna"
+seq_file = f"{dir_wk}/{folder_seq}"
+folder_gff = "data/GCF_000009045.1_ASM904v1_genomic.gff"
+gff_file= f"{dir_wk}/{folder_gff}"
```
Nice job declaring the paths at the top. I have heard that production-level code does a path.py

betteridiot replied:
Except that this is the driver function for testing purposes. There is no need to parameterize it since you will not be using these variables throughout the script.

e12chaos commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,189 @@
+        nucleotide_index = np.random.randint(0, 50-k)
+        # save the k-mer into motifs_init (list)
+        motif_picked = seqs[entry][nucleotide_index:nucleotide_index+k]
+        motifs_init.append(motif_picked)
```
comments makes it easy to follow

betteridiot replied:
But they are not always necessary if the code is written to be easy to follow.

Inline comments are for the author. Documentation is for the end-user.

zoechow24 commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,189 @@
+    
+    for entry in range(0, length_seqs):
+        # randomly select the index number within that row
+        nucleotide_index = np.random.randint(0, 50-k)
```
is 50 supposed to be the length of the sequences? If so, maybe use length_seqs instead of hardcoding. It would be more dynamic for other files.

abimckinley replied:
I agree, making this small change would make this program less tailored to the specific data that we were working with, and could be used in other scenarios.

betteridiot replied:
You should never hardcode parameters where possible. If you had iterated through it one sequence at a time, `for seq in seqs`, then you could have used `np.random.randint(0, len(seq)-k)`

zoechow24 commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,189 @@
+            max_score = max(kmer_score_list)
+            max_index = kmer_score_list.index(max_score)
+            str_update = ith_entry[max_index:max_index+k]
+            motifs_init[entry_index_picked] = str_update
```
Instead of getting the max score, try using the probabilistic selection instead. Using max may lead to the hill-climbing issue, where you get stuck at a local maximum.

Candy-Train replied:
I'm not sure if you guys had the same issue with probabilistic selection that we were having, but we ended up using relative normalization to get usable values from our scores. `random.choices` couldn't take the scores themselves for weighted values because the negative scores brought the total below zero.

betteridiot replied:
This can be solved by converting to a PFM and performing something like soft max normalization.

zoechow24 commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,189 @@
+        entry_index_picked = np.random.randint(0, length_seqs)
+        current_motif = motifs_init[:entry_index_picked] + motifs_init[entry_index_picked+1:]
+        pfm = build_pfm(current_motif, k)
+        ic = pfm_ic(pfm)
```
Are you calculating the IC according to the background PFM and using the IC to converge?

zoechow24 commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,189 @@
+        
+        if samp_times > 50:
+            window = ic_hist[-stability_window:]
+            max_change = max(window) - min(window)
```
So, to converge, are you comparing the greatest IC and the lowest IC of the last 50 indices? I'm not wrapping my head around how this shows that the IC has not changed greater than your threshold for the last x iterations. I also don't fully understand your IC because it appears that your IC is not calculated on all possible motifs, but rather only the background used to score your k-mers. Could you elaborate on your thought process in more detail?

zoechow24 commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,189 @@
+        
+        for kmer_index in range(0, 50-k):
+            motif_str = ith_entry[kmer_index:kmer_index+k]
+            motif_score = score_kmer(motif_str, pwm)
```
It looks like you forgot to score the reverse complement of your motif_str

Tiange replied:
We had the reverse complements "inverted" using the function provided

zoechow24 reviewed (commented):
Overall, it seems like you were able to get your Gibb's sampling to work even though I don't fully understand your convergence. You have a lot of nested for loops that gets hard to follow. Additional comments may help with this issue, but your pseudocode in your README definitely helped. You can also try creating support functions next time! It should make things clearer, even if the functions only have one or two lines of code, and reduce the need for nested loops.

e12chaos commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,189 @@
+            motif_score = score_kmer(motif_str, pwm)
+            kmer_score_list.append(motif_score)
+            
+            # get key(index) from the max score, update the motifs_init
+            max_score = max(kmer_score_list)
+            max_index = kmer_score_list.index(max_score)
+            str_update = ith_entry[max_index:max_index+k]
+            motifs_init[entry_index_picked] = str_update
```
I had to conduct further research on this after my initial read-through. I think what you did is called **greedy selection** The problem with using this greedy search is that it always picks the highest scoring position without exploration, causing it to get trapped in local optima where it cannot escape to find the true global best motif.

Gibbs sampling requires probabilistic selection based on score distribution, not just picking the maximum-

so i think if you replaced this section with 

`# Convert scores to probabilities
scores_array = np.array(kmer_score_list)
probs = np.exp2(scores_array)  # Convert log2 scores to probabilities
probs = probs / probs.sum()    # Normalize`

`#Probabilistically select based on scores
selected_index = np.random.choice(len(probs), p=probs)
str_update = ith_entry[selected_index:selected_index+k]
motifs_init[entry_index_picked] = str_update`

it would change your code to be a true gibbs sampling.

e12chaos commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,189 @@
+        motif_picked = seqs[entry][nucleotide_index:nucleotide_index+k]
+        motifs_init.append(motif_picked)
+    
+    for samp_times in range(0, 100):
```
This only does 100 iterations -which may have been just a testing thing, but in real senarios, this would have still been part of the burn-in phase and you'd want to do a larger sampling

Tiange replied:
Ah, yes. That's so true :) We used 100000 at first but it took too long for us to get the result (or error) so that we changed it to 100~

betteridiot replied:
But another example of hardcoding gone wrong. Parameterizing this could make playing around with the boundaries of convergence timing easier.

abimckinley commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,189 @@
+            max_change = max(window) - min(window)
+        
+            if max_change < convergence_threshold:
+                print(f"Convergence reached at iteration {j}. Max IC change: {max_change}")
```
I believe that `{j}` should actually be `{samp_times}` as "j" is undefined.

abimckinley commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,189 @@
+            kmer_score_list.append(motif_score)
+            
+            # get key(index) from the max score, update the motifs_init
+            max_score = max(kmer_score_list)
```
In addition to the below comments from Zoe and Nikaela, a small change you could make to make your program more effective would be to take the `motifs_init` updates out of the inner loop, so that it is not updating every singe time, rather than just after evaluating all the k-mers.

betteridiot replied:
I specifically stated to not use the `max` function. This is a hill-climbing approach and not probabilistic.

Candy-Train reviewed (changes requested):
I think the code is most of the way there. You just need to adjust your selection so you are choosing probabilisticly instead of using the max score.

betteridiot commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,189 @@
+    stability_window = 50  
+    convergence_threshold = 1e-5
+    
+    for entry in range(0, length_seqs):
```
This might be an oversight, but you are iterating through integers from 0 to `len(seqs)` - 1, to do...what? Get access to the index of the sequence itself within `seqs`? Aside from using functions like `enumerate`, you also don't need to do this, as you only need access to the sequence itself.

Therefore, it would almost be better to iterate through each sequence one at a time. E.g., `for seq in seqs`.

betteridiot commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,189 @@
+        # randomly select the index number within that row
+        nucleotide_index = np.random.randint(0, 50-k)
+        # save the k-mer into motifs_init (list)
+        motif_picked = seqs[entry][nucleotide_index:nucleotide_index+k]
```
Going back to my iterating through the sequences one at a time, this could have been simplified by:
```python
motif_picked = seq[nucleotide_index: nucleotide_index + k]
```

betteridiot commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,189 @@
+        current_motif = motifs_init[:entry_index_picked] + motifs_init[entry_index_picked+1:]
+        pfm = build_pfm(current_motif, k)
+        ic = pfm_ic(pfm)
+        ic_hist.append(ic)
```
I am guessing this is for conducting a rolling average.

betteridiot commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,189 @@
+        ic_hist.append(ic)
+        pwm = build_pwm(pfm)
+        
+        ith_entry = seqs[entry_index_picked]
```
Could have done this when you defined `entry_index_picked`. Also, this variable name is unique, but not as clear in its intent. Consider something like `holdout_index` versus `entry_index_picked`: what does `entry` mean here?

betteridiot commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,189 @@
+        kmer_score_list = []
+        
+        for kmer_index in range(0, 50-k):
+            motif_str = ith_entry[kmer_index:kmer_index+k]
```
You say `ith_entry`, but since `i` is never used/defined, thematically this can be confusing.

betteridiot commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,189 @@
+        ith_entry = seqs[entry_index_picked]
+        kmer_score_list = []
+        
+        for kmer_index in range(0, 50-k):
```
Again, here is where hardcoding can potentially cause problems.

betteridiot reviewed (changes requested):
Please consider the feedback and make the necessary adjustments.

mcspleen committed:
Refactor Gibbs motif finder and add softmax function

Refactor Gibbs motif finder and add softmax function.

mcspleen committed:
Revise README with AI resources and Python clarifications

Updated the README to include additional generative AI resources and clarifications on Python's yield and return statements.

mcspleen committed:
Rename function gibbs_motif_finder to GibbsMotifFinder

mcspleen committed:
Update project03.Rmd

zoechow24 commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,175 @@
+        pwm = build_pwm(pfm)
+        
+        # Track Information Content (Convergence metric)
+        current_ic = pfm_ic(pfm)
+        ic_history.append(current_ic)
```
It seems like you're only storing the IC value without comparing the old and new values. I believe your current code only converges based on iterations, which won't properly give you the global max.

Candy-Train replied:
Yeah, I think it will run through 1000 iterations  unless you do a comparison to tell it when convergence is reached.

zoechow24 reviewed (approved):
Overall, I think your code is very clean and easy to follow. The only thing I find concerning is the way you're doing convergence. As you're only converging after a set number of iterations, I don't believe you're properly comparing your IC values. While you're likely to get more internal consistency and a higher IC, I think your code will miss the global maxima and might not converge on the "true biological motif".

Candy-Train commented on project03/project03.Rmd:
```diff
@@ -48,11 +48,102 @@ from motif_ops import build_pfm, build_pwm, score_kmer, pfm_ic
+import random
 
-# Functions
+def softmax(x):
+    """Compute softmax values for each set of scores in x."""
+    e_x = np.exp(x - np.max(x))
+    return e_x / e_x.sum()
```
This looks like it should work to normalize your values and let you probabilisticly choose your next motif, which takes care of my comment from before.

Candy-Train reviewed (approved):
Looks like you got rid of the hard coded parameters and figured out a way to select your motifs without using the max, so I think this is pretty good. I dont see where you check the IC for convergence, but that would be a small fix compared to the big changes you already made. I think you've improved it a lot!

abimckinley commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,175 @@
+    for it in range(iterations):
+        
+        # Select a holdout sequence index randomly
+        holdout_idx = np.random.randint(0, num_seqs)
```
This takes care of the hard coded threshold before, much more applicable in other scenarios!

abimckinley commented on project03/project03.Rmd:
```diff
@@ -0,0 +1,175 @@
+    
+    ppfm = build_pfm(current_motifs, k)
+    return ppfm, ic_history
+  
```
The edits you made have improved this program significantly, good work!
